### PR TITLE
fix(command-dev): role based redirects

### PR DESCRIPTION
**Pull request by erezrokah**
**- Summary**

Fixes https://github.com/netlify/cli/issues/1518

**- Test plan**

Tested on this branch https://github.com/erezrokah/netlify-build-reproductions/tree/cli/issue_1518
I'll add more tests in another commit

**- Description for the changelog**

- Write a 404 response when JWT role doesn't match the route
- Set `force404` to `false` when JWT role matches the route to return a valid response

**- A picture of a cute animal (not mandatory but encouraged)**

🐱

### CODE REVIEW

File: src/utils/proxy.js [hunk 1]
```diff
@@ -171,6 +171,7 @@ const serveRedirect = async function ({ req, res, proxy, match, options }) {
         // Restore the URL if everything is correct
         if (presentedRoles.some((pr) => expectedRoles.has(pr))) {
           req.url = originalURL
+          match.force404 = false
```
Comment (erezrokah) [inline src/utils/proxy.js:174]: Not sure why `force404` is `true` when the role matches. This is returned from `netlify-redirector`.
Still looking into that one

File: src/utils/detect-server.js [hunk 2]
```diff
@@ -172,6 +172,7 @@ const serverSettings = async (devConfig, flags, projectDir, log) => {
     throw new Error(`Could not acquire required "port": ${triedPort}`)
   }
 
+  settings.jwtSecret = devConfig.jwtSecret || 'secret'
```
Comment (erezrokah) [inline src/utils/detect-server.js:175]: The option to configure a custom `jwtSecret` was missing

File: src/utils/proxy.js [hunk 3]
```diff
@@ -187,7 +187,8 @@ const serveRedirect = async function ({ req, res, proxy, match, options }) {
   if (staticFile) req.url = staticFile + reqUrl.search
   if (match.force404) {
     res.writeHead(404)
-    return render404(options.publicFolder)
```
Comment (erezrokah) [inline src/utils/proxy.js:190]: This was causing requests to Netlify Dev to hang

File: src/utils/rules-proxy.js [hunk 4]
```diff
@@ -76,8 +76,8 @@ const createRewriter = async function ({ distDir, projectDir, jwtSecret, jwtRole
 
     if (rules.length !== 0) {
       return (matcher = await redirector.parseJSON(JSON.stringify(rules), {
-        jwtSecret: jwtSecret || 'secret',
-        jwtRole: jwtRole || 'app_metadata.authorization.roles',
```
Comment (erezrokah) [inline src/utils/rules-proxy.js:80]: Arguments should be `jwtSecret` and `jwtRoleClaim`.
Default values are already assigned when reading the configuration

File: src/utils/rules-proxy.js [hunk 5]
```diff
@@ -109,6 +109,9 @@ const createRewriter = async function ({ distDir, projectDir, jwtSecret, jwtRole
       path: reqUrl.pathname,
       query: reqUrl.search.slice(1),
       headers,
+      // passing an empty object for cookies is required due to a bug in netlify-redirector
+      // actual cookie values are retrieved via getCookie
+      cookies: {},
```
Comment (erezrokah) [inline src/utils/rules-proxy.js:114]: I submitted a PR to `netlify-redirector` to fix this

**Review by ehmicky — APPROVED** [review]
:+1:

**Pull request merged**